File: vm/minimal/nga.zig
// Copyright (c) awawawawawa
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the copyright
// notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
//
// Minimal Nga virtual machine for RETRO, implemented in Zig.
// Build with: zig build-exe nga.zig
const std = @import("std");

const Cell = i32;
const IMAGE_SIZE: usize = 65536;
const ADDRESSES: usize = 256;
const STACK_DEPTH: usize = 256;
const CELL_MIN: Cell = std.math.minInt(Cell) + 1;
const CELL_MAX: Cell = std.math.maxInt(Cell) - 1;

const NgaVm = struct {
    sp: usize = 0,
    rp: usize = 0,
    ip: Cell = 0,
    data: [STACK_DEPTH]Cell = [_]Cell{0} ** STACK_DEPTH,
    address: [ADDRESSES]Cell = [_]Cell{0} ** ADDRESSES,
    memory: [IMAGE_SIZE + 1]Cell = [_]Cell{0} ** (IMAGE_SIZE + 1),

    fn halt(self: *NgaVm) void {
        self.ip = IMAGE_SIZE;
    }

    fn validMemoryAddress(value: Cell) bool {
        return value >= 0 and value <= IMAGE_SIZE;
    }

    fn canPushData(self: *const NgaVm) bool {
        return self.sp < STACK_DEPTH - 1;
    }

    fn canPopData(self: *const NgaVm) bool {
        return self.sp > 0;
    }

    fn canPushAddress(self: *const NgaVm) bool {
        return self.rp < ADDRESSES - 1;
    }

    fn canPopAddress(self: *const NgaVm) bool {
        return self.rp > 0;
    }

    fn tos(self: *const NgaVm) Cell {
        return self.data[self.sp];
    }

    fn nos(self: *const NgaVm) Cell {
        return self.data[self.sp - 1];
    }

    fn tors(self: *const NgaVm) Cell {
        return self.address[self.rp];
    }

    fn setTos(self: *NgaVm, value: Cell) void {
        self.data[self.sp] = value;
    }

    fn setNos(self: *NgaVm, value: Cell) void {
        self.data[self.sp - 1] = value;
    }

    fn setTors(self: *NgaVm, value: Cell) void {
        self.address[self.rp] = value;
    }

    fn stackPush(self: *NgaVm, value: Cell) void {
        if (!self.canPushData()) {
            self.halt();
            return;
        }
        self.sp += 1;
        self.setTos(value);
    }

    fn stackPop(self: *NgaVm) Cell {
        if (!self.canPopData()) {
            self.halt();
            return 0;
        }
        const value = self.tos();
        self.data[self.sp] = 0;
        self.sp -= 1;
        return value;
    }

    fn loadImage(self: *NgaVm, io: std.Io, image_name: []const u8) !void {
        var file = try std.Io.Dir.cwd().openFile(io, image_name, .{});
        defer file.close(io);
        var reader = file.reader(io, &.{});

        var bytes: [4]u8 = undefined;
        var index: usize = 0;
        while (index < IMAGE_SIZE) : (index += 1) {
            reader.interface.readSliceAll(&bytes) catch |err| switch (err) {
                error.EndOfStream => break,
                else => return err,
            };
            self.memory[index] = std.mem.readInt(Cell, &bytes, .little);
        }
    }

    fn execute(self: *NgaVm, io: std.Io, cell: Cell) void {
        self.rp = 1;
        self.ip = cell;
        while (self.ip >= 0 and self.ip < IMAGE_SIZE) {
            const opcode = self.memory[@intCast(self.ip)];
            self.processOpcodeBundle(io, opcode);
            self.ip += 1;
            if (self.rp == 0) self.ip = IMAGE_SIZE;
        }
    }

    fn instLi(self: *NgaVm) void {
        if (self.ip >= IMAGE_SIZE - 1 or !self.canPushData()) return self.halt();
        self.ip += 1;
        self.stackPush(self.memory[@intCast(self.ip)]);
    }

    fn instDu(self: *NgaVm) void {
        if (!self.canPopData()) return self.halt();
        self.stackPush(self.tos());
    }

    fn instDr(self: *NgaVm) void {
        _ = self.stackPop();
    }

    fn instSw(self: *NgaVm) void {
        if (self.sp < 2) return self.halt();
        const value = self.tos();
        self.setTos(self.nos());
        self.setNos(value);
    }

    fn instPu(self: *NgaVm) void {
        if (!self.canPushAddress() or !self.canPopData()) return self.halt();
        self.rp += 1;
        self.setTors(self.tos());
        self.instDr();
    }

    fn instPo(self: *NgaVm) void {
        if (!self.canPopAddress() or !self.canPushData()) return self.halt();
        self.stackPush(self.tors());
        self.rp -= 1;
    }

    fn instJu(self: *NgaVm) void {
        if (!self.canPopData() or self.tos() < 0 or self.tos() >= IMAGE_SIZE) return self.halt();
        self.ip = self.tos() - 1;
        self.instDr();
    }

    fn instCa(self: *NgaVm) void {
        if (!self.canPushAddress() or !self.canPopData() or self.tos() < 0 or self.tos() >= IMAGE_SIZE) return self.halt();
        self.rp += 1;
        self.setTors(self.ip);
        self.ip = self.tos() - 1;
        self.instDr();
    }

    fn instCc(self: *NgaVm) void {
        if (self.sp < 2) return self.halt();
        const target = self.stackPop();
        const flag = self.stackPop();
        if (flag != 0) {
            if (!self.canPushAddress() or target < 0 or target >= IMAGE_SIZE) return self.halt();
            self.rp += 1;
            self.setTors(self.ip);
            self.ip = target - 1;
        }
    }

    fn instRe(self: *NgaVm) void {
        if (!self.canPopAddress() or self.tors() < 0 or self.tors() >= IMAGE_SIZE) return self.halt();
        self.ip = self.tors();
        self.rp -= 1;
    }

    fn compare(self: *NgaVm, result: bool) void {
        if (self.sp < 2) return self.halt();
        self.setNos(if (result) -1 else 0);
        self.instDr();
    }

    fn instFe(self: *NgaVm) void {
        if (!self.canPopData()) return self.halt();
        self.setTos(switch (self.tos()) {
            -1 => @as(Cell, @intCast(self.sp - 1)),
            -2 => @as(Cell, @intCast(self.rp)),
            -3 => IMAGE_SIZE,
            -4 => CELL_MIN,
            -5 => CELL_MAX,
            else => |address| if (validMemoryAddress(address)) self.memory[@intCast(address)] else return self.halt(),
        });
    }

    fn instSt(self: *NgaVm) void {
        if (self.sp < 2) return self.halt();
        const address = self.tos();
        if (!validMemoryAddress(address)) return self.halt();
        self.memory[@intCast(address)] = self.nos();
        self.instDr();
        self.instDr();
    }

    fn instDi(self: *NgaVm) void {
        if (self.sp < 2) return self.halt();
        const divisor = self.tos();
        const dividend = self.nos();
        if (divisor == 0 or (divisor == -1 and dividend == std.math.minInt(Cell))) return self.halt();
        self.setTos(@divTrunc(dividend, divisor));
        self.setNos(@rem(dividend, divisor));
    }

    fn instSh(self: *NgaVm) void {
        if (self.sp < 2) return self.halt();
        const shift = self.tos();
        const value = self.nos();
        if (shift < -31 or shift > 31) return self.halt();
        const amount: u5 = @intCast(if (shift < 0) -shift else shift);
        const bits: u32 = @bitCast(value);
        const shifted: Cell = if (shift < 0) @bitCast(bits << amount) else value >> amount;
        self.setNos(shifted);
        self.instDr();
    }

    fn instIi(self: *NgaVm, io: std.Io) void {
        if (!self.canPopData()) return self.halt();
        switch (self.tos()) {
            0 => {
                self.instDr();
                const value = self.stackPop();
                if (self.ip == IMAGE_SIZE) return;
                var writer = std.Io.File.stdout().writer(io, &.{});
                writer.interface.writeAll(&[_]u8{@truncate(@as(u32, @bitCast(value)))}) catch self.halt();
            },
            1 => {
                var byte: [1]u8 = undefined;
                var reader = std.Io.File.stdin().reader(io, &.{});
                const count = reader.interface.readSliceShort(&byte) catch return self.halt();
                if (count == 0) return self.halt();
                self.instDr();
                self.stackPush(byte[0]);
            },
            else => self.instDr(),
        }
    }

    fn instruction(self: *NgaVm, io: std.Io, opcode: u8) void {
        switch (opcode) {
            0 => {},
            1 => self.instLi(),
            2 => self.instDu(),
            3 => self.instDr(),
            4 => self.instSw(),
            5 => self.instPu(),
            6 => self.instPo(),
            7 => self.instJu(),
            8 => self.instCa(),
            9 => self.instCc(),
            10 => self.instRe(),
            11 => if (self.sp < 2) self.halt() else self.compare(self.nos() == self.tos()),
            12 => if (self.sp < 2) self.halt() else self.compare(self.nos() != self.tos()),
            13 => if (self.sp < 2) self.halt() else self.compare(self.nos() < self.tos()),
            14 => if (self.sp < 2) self.halt() else self.compare(self.nos() > self.tos()),
            15 => self.instFe(),
            16 => self.instSt(),
            17 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() +% self.tos());
                    self.instDr();
                }
            },
            18 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() -% self.tos());
                    self.instDr();
                }
            },
            19 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() *% self.tos());
                    self.instDr();
                }
            },
            20 => self.instDi(),
            21 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() & self.tos());
                    self.instDr();
                }
            },
            22 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() | self.tos());
                    self.instDr();
                }
            },
            23 => {
                if (self.sp < 2) self.halt() else {
                    self.setNos(self.nos() ^ self.tos());
                    self.instDr();
                }
            },
            24 => self.instSh(),
            25 => if (self.canPopData() and self.tos() == 0) {
                if (!self.canPopAddress() or self.tors() < 0 or self.tors() >= IMAGE_SIZE) self.halt() else {
                    self.instDr();
                    self.ip = self.tors();
                    self.rp -= 1;
                }
            },
            26 => self.halt(),
            27 => self.stackPush(2),
            28 => if (!self.canPopData()) self.halt() else if (self.tos() == 0) {
                self.instDr();
                self.stackPush(0);
                self.stackPush(0);
            } else if (self.tos() == 1) {
                self.instDr();
                self.stackPush(1);
                self.stackPush(1);
            },
            29 => self.instIi(io),
            else => self.halt(),
        }
    }

    fn processOpcodeBundle(self: *NgaVm, io: std.Io, opcode: Cell) void {
        const bits: u32 = @bitCast(opcode);
        inline for ([_]u5{ 0, 8, 16, 24 }) |shift| {
            self.instruction(io, @truncate(bits >> shift));
        }
    }
};

pub fn main(init: std.process.Init) void {
    var vm = NgaVm{};
    vm.loadImage(init.io, "ngaImage") catch return;
    vm.execute(init.io, 0);
}

File: vm/minimal/nga.py
#!/usr/bin/env python3
"""Minimal Nga virtual machine for RETRO.

Run with: python3 nga.py
"""

import struct
import sys


IMAGE_SIZE = 65536
ADDRESSES = 256
STACK_DEPTH = 256
CELL_MIN = -2147483647
CELL_MAX = 2147483646
CELL_MASK = 0xFFFFFFFF


def as_cell(value):
    """Convert value to a signed, 32-bit Nga cell."""
    value &= CELL_MASK
    return value if value < 0x80000000 else value - 0x100000000


class NgaVm:
    def __init__(self):
        self.sp = 0
        self.rp = 0
        self.ip = 0
        self.data = [0] * STACK_DEPTH
        self.address = [0] * ADDRESSES
        self.memory = [0] * (IMAGE_SIZE + 1)

    def halt(self):
        self.ip = IMAGE_SIZE

    @staticmethod
    def valid_memory_address(value):
        return 0 <= value <= IMAGE_SIZE

    def can_push_data(self):
        return self.sp < STACK_DEPTH - 1

    def can_pop_data(self):
        return self.sp > 0

    def can_push_address(self):
        return self.rp < ADDRESSES - 1

    def can_pop_address(self):
        return self.rp > 0

    def tos(self):
        return self.data[self.sp]

    def nos(self):
        return self.data[self.sp - 1]

    def tors(self):
        return self.address[self.rp]

    def stack_push(self, value):
        if not self.can_push_data():
            self.halt()
            return
        self.sp += 1
        self.data[self.sp] = as_cell(value)

    def stack_pop(self):
        if not self.can_pop_data():
            self.halt()
            return 0
        value = self.data[self.sp]
        self.data[self.sp] = 0
        self.sp -= 1
        return value

    def load_image(self, image_name):
        with open(image_name, "rb") as image:
            for index in range(IMAGE_SIZE):
                cell = image.read(4)
                if not cell:
                    return
                if len(cell) != 4:
                    raise ValueError("image is not a sequence of 32-bit cells")
                self.memory[index] = struct.unpack("<i", cell)[0]

    def execute(self, cell):
        self.rp = 1
        self.ip = cell
        while 0 <= self.ip < IMAGE_SIZE:
            self.process_opcode_bundle(self.memory[self.ip])
            self.ip += 1
            if self.rp == 0:
                self.ip = IMAGE_SIZE

    def inst_li(self):
        if self.ip >= IMAGE_SIZE - 1 or not self.can_push_data():
            self.halt()
            return
        self.ip += 1
        self.stack_push(self.memory[self.ip])

    def inst_du(self):
        if not self.can_pop_data():
            self.halt()
            return
        self.stack_push(self.tos())

    def inst_dr(self):
        self.stack_pop()

    def inst_sw(self):
        if self.sp < 2:
            self.halt()
            return
        self.data[self.sp], self.data[self.sp - 1] = self.nos(), self.tos()

    def inst_pu(self):
        if not self.can_push_address() or not self.can_pop_data():
            self.halt()
            return
        self.rp += 1
        self.address[self.rp] = self.tos()
        self.inst_dr()

    def inst_po(self):
        if not self.can_pop_address() or not self.can_push_data():
            self.halt()
            return
        self.stack_push(self.tors())
        self.rp -= 1

    def inst_ju(self):
        if not self.can_pop_data() or not 0 <= self.tos() < IMAGE_SIZE:
            self.halt()
            return
        self.ip = self.tos() - 1
        self.inst_dr()

    def inst_ca(self):
        if not self.can_push_address() or not self.can_pop_data() or not 0 <= self.tos() < IMAGE_SIZE:
            self.halt()
            return
        self.rp += 1
        self.address[self.rp] = self.ip
        self.ip = self.tos() - 1
        self.inst_dr()

    def inst_cc(self):
        if self.sp < 2:
            self.halt()
            return
        target = self.stack_pop()
        flag = self.stack_pop()
        if flag:
            if not self.can_push_address() or not 0 <= target < IMAGE_SIZE:
                self.halt()
                return
            self.rp += 1
            self.address[self.rp] = self.ip
            self.ip = target - 1

    def inst_re(self):
        if not self.can_pop_address() or not 0 <= self.tors() < IMAGE_SIZE:
            self.halt()
            return
        self.ip = self.tors()
        self.rp -= 1

    def compare(self, result):
        if self.sp < 2:
            self.halt()
            return
        self.data[self.sp - 1] = -1 if result else 0
        self.inst_dr()

    def inst_fe(self):
        if not self.can_pop_data():
            self.halt()
            return
        value = self.tos()
        if value == -1:
            self.data[self.sp] = self.sp - 1
        elif value == -2:
            self.data[self.sp] = self.rp
        elif value == -3:
            self.data[self.sp] = IMAGE_SIZE
        elif value == -4:
            self.data[self.sp] = CELL_MIN
        elif value == -5:
            self.data[self.sp] = CELL_MAX
        elif self.valid_memory_address(value):
            self.data[self.sp] = self.memory[value]
        else:
            self.halt()

    def inst_st(self):
        if self.sp < 2:
            self.halt()
            return
        address = self.tos()
        if not self.valid_memory_address(address):
            self.halt()
            return
        self.memory[address] = self.nos()
        self.inst_dr()
        self.inst_dr()

    def inst_di(self):
        if self.sp < 2:
            self.halt()
            return
        divisor = self.tos()
        dividend = self.nos()
        if divisor == 0 or (divisor == -1 and dividend == -2147483648):
            self.halt()
            return
        quotient = abs(dividend) // abs(divisor)
        if (dividend < 0) != (divisor < 0):
            quotient = -quotient
        self.data[self.sp] = quotient
        self.data[self.sp - 1] = dividend - quotient * divisor

    def inst_sh(self):
        if self.sp < 2:
            self.halt()
            return
        shift = self.tos()
        value = self.nos()
        if not -31 <= shift <= 31:
            self.halt()
            return
        if shift < 0:
            result = as_cell((value & CELL_MASK) << -shift)
        else:
            result = value >> shift
        self.data[self.sp - 1] = result
        self.inst_dr()

    def inst_zr(self):
        if not self.can_pop_data():
            self.halt()
            return
        if self.tos() != 0:
            return
        if not self.can_pop_address() or not 0 <= self.tors() < IMAGE_SIZE:
            self.halt()
            return
        self.inst_dr()
        self.ip = self.tors()
        self.rp -= 1

    def inst_iq(self):
        if not self.can_pop_data():
            self.halt()
        elif self.tos() == 0:
            self.inst_dr()
            self.stack_push(0)
            self.stack_push(0)
        elif self.tos() == 1:
            self.inst_dr()
            self.stack_push(1)
            self.stack_push(1)

    def inst_ii(self):
        if not self.can_pop_data():
            self.halt()
        elif self.tos() == 0:
            self.inst_dr()
            value = self.stack_pop()
            if self.ip != IMAGE_SIZE:
                sys.stdout.buffer.write(bytes((value & 0xFF,)))
                sys.stdout.buffer.flush()
        elif self.tos() == 1:
            value = sys.stdin.buffer.read(1)
            if not value:
                self.halt()
                return
            self.inst_dr()
            self.stack_push(value[0])
        else:
            self.inst_dr()

    def instruction(self, opcode):
        if opcode == 0:
            return
        if opcode == 1:
            self.inst_li()
        elif opcode == 2:
            self.inst_du()
        elif opcode == 3:
            self.inst_dr()
        elif opcode == 4:
            self.inst_sw()
        elif opcode == 5:
            self.inst_pu()
        elif opcode == 6:
            self.inst_po()
        elif opcode == 7:
            self.inst_ju()
        elif opcode == 8:
            self.inst_ca()
        elif opcode == 9:
            self.inst_cc()
        elif opcode == 10:
            self.inst_re()
        elif opcode == 11:
            self.compare(self.nos() == self.tos()) if self.sp >= 2 else self.halt()
        elif opcode == 12:
            self.compare(self.nos() != self.tos()) if self.sp >= 2 else self.halt()
        elif opcode == 13:
            self.compare(self.nos() < self.tos()) if self.sp >= 2 else self.halt()
        elif opcode == 14:
            self.compare(self.nos() > self.tos()) if self.sp >= 2 else self.halt()
        elif opcode == 15:
            self.inst_fe()
        elif opcode == 16:
            self.inst_st()
        elif opcode == 17:
            self.compare_binary(lambda a, b: as_cell(a + b))
        elif opcode == 18:
            self.compare_binary(lambda a, b: as_cell(a - b))
        elif opcode == 19:
            self.compare_binary(lambda a, b: as_cell(a * b))
        elif opcode == 20:
            self.inst_di()
        elif opcode == 21:
            self.compare_binary(lambda a, b: a & b)
        elif opcode == 22:
            self.compare_binary(lambda a, b: a | b)
        elif opcode == 23:
            self.compare_binary(lambda a, b: a ^ b)
        elif opcode == 24:
            self.inst_sh()
        elif opcode == 25:
            self.inst_zr()
        elif opcode == 26:
            self.halt()
        elif opcode == 27:
            self.stack_push(2)
        elif opcode == 28:
            self.inst_iq()
        elif opcode == 29:
            self.inst_ii()
        else:
            self.halt()

    def compare_binary(self, operation):
        if self.sp < 2:
            self.halt()
            return
        self.data[self.sp - 1] = operation(self.nos(), self.tos())
        self.inst_dr()

    def process_opcode_bundle(self, opcode):
        opcode &= CELL_MASK
        for shift in (0, 8, 16, 24):
            self.instruction((opcode >> shift) & 0xFF)


def main():
    vm = NgaVm()
    try:
        vm.load_image("ngaImage")
    except (OSError, ValueError):
        return
    vm.execute(0)


if __name__ == "__main__":
    main()
